### PR TITLE
Create outline of GUI

### PR DESCRIPTION
**Pull request by EYangster**
* Add menu button for CSV export
* Add horizontal offsetting and scaling (applies to both channels)
* Add vertical offsetting and scaling (applies to selected channel(s))
* Add reset and trigger buttons

Updated dependencies as newer feature was needed (`Button::min_size()` was private)

Signed-off-by: Eric Yang <e.yang@queensu.ca>

### CODE REVIEW

**Comment by EYangster** [comment]
Purpose of `saved_v_offset` and `saved_v_scale` (TBD):
When a channel is turned on, we save the offset and scale of the sliders so that moving them will update the channel's offset/scale by adding the difference of the current slider offset/scale and the saved offset/scale. When the channel is deselected we save to the channel's own offset/scale parameters.

**Comment by EYangster** [comment]
image

Reset button resets all widgets in "Channel scaling"
Trigger button toggles between "Start" and "Stop" text

File: software/teachee-desktop/src/app.rs
Comment (ethanmpeterson) [inline software/teachee-desktop/src/app.rs:33]: Pardon me if I am misunderstanding but everything in this struct is protected by the mutex correct? If that is the case I think we can pull a lot of information out of this struct, the USB manager thread does not need to know anything about our scaling our how we process the samples. However, for trigger control, we will need some shared info between the threads. Maybe pull out everything except the trigger control? The rest can be local I think. Even the offsets can probably be pulled out. That said, reducing the amount of data in this struct may not even really have any affect on performance. So I leave it to you guys.

File: software/teachee-desktop/src/app.rs
Comment (ethanmpeterson) [inline software/teachee-desktop/src/app.rs:251]: I cannot comment much on the UI code but I like the way it looks. One extension that I think would be really handy is to also have a text box available for entering the offsets and scales rather than sliders. For instance, if you are trying to scope a signal at a known frequency, it would be really helpful to just type "1" s/div or V/div into a text box to set the scale. The sliders should also be kept though as they replicate a physical knob you see on other scopes.

File: software/teachee-desktop/Cargo.toml
Comment (ethanmpeterson) [inline software/teachee-desktop/Cargo.toml:9]: Was this version upgrade automatic? or did you do that get some kind of new feature?

File: software/teachee-desktop/src/app.rs
Comment (EYangster) [reply software/teachee-desktop/src/app.rs:33]: This is what I proposed in https://github.com/teachee-capstone/teachee/pull/31. Perhaps it should be reconsidered? @thetimmorland

File: software/teachee-desktop/src/app.rs
Comment (EYangster) [reply software/teachee-desktop/src/app.rs:251]: What would the units for offset be? Do we need to manually enter offsets? I think scale is enough based on other oscilloscopes.

Comment (EYangster) [reply software/teachee-desktop/src/app.rs:251]: I will add editable text boxes that display the s/div and channel 1 and 2 V/div and perhaps others

File: software/teachee-desktop/Cargo.toml
Comment (EYangster) [reply software/teachee-desktop/Cargo.toml:9]: I needed a function that was private in the old version but public in the current so I updated (see PR description)

File: software/teachee-desktop/src/app.rs
Comment (thetimmorland) [inline software/teachee-desktop/src/app.rs]: This is basically hardcoding `self = Default();` for a subset of the fields on `self`. It might make more sense to split these fields into a new struct so we can reset them all at once using something like `app.scope_ui = Default();`.

File: software/teachee-desktop/src/app.rs
Comment (thetimmorland) [inline software/teachee-desktop/src/app.rs]: While you are working with this code could you also change these fields to be an array?

File: software/teachee-desktop/src/app.rs
Comment (thetimmorland) [inline software/teachee-desktop/src/app.rs]: Check out https://doc.rust-lang.org/std/macro.todo.html. Not important in this case, but it's handy since it makes unfinished code typecheck. Though it will crash the program if the button is clicked.

File: software/teachee-desktop/src/app.rs
Comment (thetimmorland) [reply software/teachee-desktop/src/app.rs:37]: I'd like to leave it as is for now. If / when we run into performance issues this the first thing to try though.

Comment (thetimmorland) [reply software/teachee-desktop/src/app.rs:33]: I'd like to leave it as is for now. If we run into performance problems we should definitely try the other approach.

File: software/teachee-desktop/src/app.rs
Comment (EYangster) [reply software/teachee-desktop/src/app.rs]: Can't do `*self = Self::default();` since we're mutably borrowing self's fields twice (one at `let Self {} = self;` and the other when setting to default). So I assume if we destructure then we can't access through self.

Comment (EYangster) [reply software/teachee-desktop/src/app.rs]: See latest commit. We either destructure everything but manually assign in reset or keep UIControls but have to litter the code with `&mut ui_controls.<field>`

File: software/teachee-desktop/src/app.rs
Comment (EYangster) [reply software/teachee-desktop/src/app.rs]: I think it's best we do this in a new PR since this one has enough changes already. I will make a new PR after this is merged to clean up the old plot code

**Review by thetimmorland — APPROVED** [review]
Looks good to me. I'd still like to try and put channel controls into a vector at some point.

If we did that we could iterate over the vector and draw a separate controls section per channel. This would allow us to model adding a FFT to the plot as "adding" a new channel to the vector, so as the user toggles FFT the relavent controls appear and disappear.

**Pull request merged**
